### ISSUE TITLE
retrieve_data cannot be filtered or stopped before it has fetched everything

### ISSUE THREAD

**Issue by 8h2a** [josegonzalez/python-github-backup#119]
I tried to (incrementally) backup a repository with thousands of pull requests and noticed that it kept retrieving the same pull requests that were already backed up. (The same problem probably also applies for issues/gists/...?)

The problem is that `retrieve_data` cannot be filtered or stopped until is has fetched _all_ the data and the function/API does not allow to filter based on the `updated_at` date.

I propose to change `retrieve_data` to make it a generator instead of returning a list.
Below is a prototype patch that fixes the issue for pull requests. I renamed `retrieve_data` to `retrieve_data_gen`, and added another `retrieve_data` function which collects everything into a list to provide the same old functionality which is required by many other parts. Then I used `retrieve_data_gen` and stop when we reached pull requests that are older than `args.since`. This works because the API call is already sorted by date in descending order.

Please let me know what you think about this!
Should I prepare a pull request for this?
This patch should also help with #76.

```diff
diff --git a/bin/github-backup b/bin/github-backup
index 2dd27c8..87fa565 100755
--- a/bin/github-backup
+++ b/bin/github-backup
@@ -406,12 +406,11 @@ def get_github_repo_url(args, repository):
     return repo_url
 
 
-def retrieve_data(args, template, query_args=None, single_request=False):
+def retrieve_data_gen(args, template, query_args=None, single_request=False):
     auth = get_auth(args)
     query_args = get_query_args(query_args)
     per_page = 100
     page = 0
-    data = []
 
     while True:
         page = page + 1
@@ -438,11 +437,12 @@ def retrieve_data(args, template, query_args=None, single_request=False):
         response = json.loads(r.read().decode('utf-8'))
         if len(errors) == 0:
             if type(response) == list:
-                data.extend(response)
+                for resp in response:
+                    yield resp
                 if len(response) < per_page:
                     break
             elif type(response) == dict and single_request:
-                data.append(response)
+                yield response
 
         if len(errors) > 0:
             log_error(errors)
@@ -450,8 +450,8 @@ def retrieve_data(args, template, query_args=None, single_request=False):
         if single_request:
             break
 
-    return data
-
+def retrieve_data(args, template, query_args=None, single_request=False):
+    return list(retrieve_data_gen(args, template, query_args, single_request))
 
 def get_query_args(query_args=None):
     if not query_args:
@@ -836,18 +836,21 @@ def backup_pulls(args, repo_cwd, repository, repos_template):
         pull_states = ['open', 'closed']
         for pull_state in pull_states:
             query_args['state'] = pull_state
-            # It'd be nice to be able to apply the args.since filter here...
-            _pulls = retrieve_data(args,
+            _pulls = retrieve_data_gen(args,
                                    _pulls_template,
                                    query_args=query_args)
             for pull in _pulls:
+                if args.since and pull['updated_at'] < args.since:
+                    break
                 if not args.since or pull['updated_at'] >= args.since:
                     pulls[pull['number']] = pull
     else:
-        _pulls = retrieve_data(args,
+        _pulls = retrieve_data_gen(args,
                                _pulls_template,
                                query_args=query_args)
         for pull in _pulls:
+            if args.since and pull['updated_at'] < args.since:
+                break
             if not args.since or pull['updated_at'] >= args.since:
                 pulls[pull['number']] = retrieve_data(
                     args,
```

**Comment by josegonzalez** [comment]
Pull requests welcome.

**Comment by josegonzalez** [comment]
Released in 0.25.0.
